[BEAM-4076] Import Schema branch into master

reuvenlax opened the pull request:
This is the import of the schema branch into master. This implements basic end-to-end support of Schemas, along with automatic inference of schemas.

These APIs are not yet final, and are all marked Experimental.

R: @akedin

reuvenlax commented:
Spurious test failures are fixed. Dataflow tests are expected to fail until the Dataflow worker has been updated.

kennknowles commented:
This seems to have a bunch of commits that have already been moved and also some that are sort of intermediate fixups. Can you squash & curate a bit?

reuvenlax commented:
Squashed a bunch of commits.

Unfortunately, I'm not sure what to do about the commits that were already
moved. I don't think they are all identical to the ones that ended up in
master (as there was some editing of them in another branch before merging
to master), rather they are roughly similar to what's in master; later
commits in this branch bring it back in line with master.

On Mon, Jun 4, 2018 at 4:33 PM Kenn Knowles <notifications@github.com>
wrote:

> This seems to have a bunch of commits that have already been moved and
> also some that are sort of intermediate fixups. Can you squash & curate a
> bit?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/beam/pull/5545#issuecomment-394356109>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AUGE1cfoGJ0e6_2ulRuv0eccERXf8uOUks5t5TdEgaJpZM4UYOHO>
> .
>

reuvenlax commented:
@akedin friendly ping. any update on this review?

reuvenlax commented:
@chamikaramj has agreed to help review the Beam-core parts of this PR. That is largely the following files:
DoFnRunners.java
SimpleDoFnRunner.java
ParDoEvaluator.java
DataflowPipelineTranslator.java
Pipeline.java
Create.java
DoFn.java
DoFnOutputReceivers.java
ParDo.java
ByteBuddyDoFnInvokerFactory.java
DoFnInvoker.java
DoFnSignature.java
DoFnSignatures.java
DoFnInfo.java
PCollection.java

chamikaramj commented on sdks/java/core/src/main/java/org/apache/beam/sdk/values/PCollection.java:
```diff
@@ -276,6 +299,18 @@ public String getName() {
   }
 
+  /**
+   * Sets a {@link Schema} on this {@link PCollection}. This is a wrapper around
```
Why reveal implementation details here ?

chamikaramj commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/reflect/DoFnSignatures.java:
```diff
@@ -41,9 +41,11 @@
 import javax.annotation.Nullable;
 import org.apache.beam.sdk.coders.Coder;
 import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.schemas.FieldAccessDescriptor;
```
Add unit tests to DoFnSignaturesTest to cover updates here ?

reuvenlax replied:
Done

chamikaramj commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ParDo.java:
```diff
@@ -479,6 +483,43 @@ private static void validateStateApplicableForInput(
+      // In this case, we expect to have a FieldAccessDescriptor defined in the class.
+      FieldAccessDeclaration fieldAccessDeclaration = fieldAccessDeclarations.get(id);
+      checkArgument(fieldAccessDeclaration != null,
+          "No FieldAccessDescriptor defined with id", id);
```
FieldAccessDeclaration ?

chamikaramj commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/ParDo.java:
```diff
@@ -778,6 +820,18 @@ public PCollectionTuple expand(PCollection<? extends InputT> input) {
 
+      DoFnSignature.ProcessElementMethod processElementMethod = signature.processElement();
+      RowParameter rowParameter = processElementMethod.getRowParameter();
+      // Can only as for a Row if a Schema was specified!
```
"ask for a"

chamikaramj commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFn.java:
```diff
@@ -376,7 +377,15 @@ public Duration getAllowedTimestampSkew() {
+     * <p>The {@link PCollection} representing this tag must have a schema registered in order to
+     * call this function.
+     */
+    <T> OutputReceiver<Row> getRowReceiver(TupleTag<T> tag);
```
@Experimental(Kind.SCHEMAS) ?

chamikaramj commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java:
```diff
@@ -93,6 +100,17 @@
+  @Nullable
+  private final SchemaCoder<InputT> schemaCoder;
+
+  @Nullable final SchemaCoder<OutputT> mainOutputSchemaCoder;
```
Why is mainOutput special in this case ? Don't we need schema coders for other outputs ?

reuvenlax replied:
It's used for the default output() method, which always produces to the main output.

reuvenlax replied:
It's used for the default output() method, which always produces to the main output.

chamikaramj commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java:
```diff
@@ -101,19 +119,64 @@ public SimpleDoFnRunner(
         Sets.newHashSet(FluentIterable.<TupleTag<?>>of(mainOutputTag).append(additionalOutputTags));
     this.stepContext = stepContext;
 
+    // Currently we only support a single FieldAccess on a processElement. We should decide
```
s/FieldAccess/FieldAccessDescriptor/g

chamikaramj commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java:
```diff
@@ -101,19 +119,64 @@ public SimpleDoFnRunner(
+    FieldAccessDescriptor fieldAccessDescriptor = null;
+    if (rowParameter != null) {
+      checkArgument(schemaCoder != null,
+          "Cannot access object as a row if the input PCollection does not have a schema ."
```
Does "no schema coder" ==> "no schema" or do we have to talk about not having the schema coder defined in this exception message ?

reuvenlax replied:
Schema is currently represented under the covers by SchemaCoder. So no schema coder === no schema.

reuvenlax replied:
Schema is currently represented under the covers by SchemaCoder. So no schema coder === no schema.

chamikaramj commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java:
```diff
@@ -101,19 +119,64 @@ public SimpleDoFnRunner(
+        FieldAccessDeclaration fieldAccessDeclaration =
+            doFnSignature.fieldAccessDeclarations().get(id);
+        checkArgument(fieldAccessDeclaration != null,
+            "No FieldAccessDescriptor defined with id", id);
```
s/FieldAccessDescriptor/FieldAccessDeclaration/ ?

reuvenlax commented:
@chamikaramj thank you. addressed your comments.

chamikaramj commented:
Thanks. LGTM for updates to following classes.

DoFnRunners.java
SimpleDoFnRunner.java
ParDoEvaluator.java
DataflowPipelineTranslator.java
Pipeline.java
Create.java
DoFn.java
DoFnOutputReceivers.java
ParDo.java
ByteBuddyDoFnInvokerFactory.java
DoFnInvoker.java
DoFnSignature.java
DoFnSignatures.java
DoFnInfo.java
PCollection.java

kennknowles commented:
We have turned on autoformatting of the codebase, which causes small conflicts across the board. You can probably safely rebase and just keep your changes. Like this:

```
$ git rebase
... see some conflicts
$ git diff
... confirmed that the conflicts are just autoformatting
... so we can just keep our changes are do our own autoformat
$ git checkout --theirs --
$ git add -u
$ git rebase --continue
$ ./gradlew spotlessJavaApply
```

Please ping me if you run into any difficulty.

reuvenlax commented:
This command sequence doesn't actually work :)

git checkout --theirs --

fatal: '--ours/--theirs' cannot be used with switching branches


On Thu, Jun 28, 2018 at 9:24 AM Kenn Knowles <notifications@github.com>
wrote:

> We have turned on autoformatting of the codebase, which causes small
> conflicts across the board. You can probably safely rebase and just keep
> your changes. Like this:
>
> $ git rebase
> ... see some conflicts
> $ git diff
> ... confirmed that the conflicts are just autoformatting
> ... so we can just keep our changes are do our own autoformat
> $ git checkout --theirs --
> $ git add -u
> $ git rebase --continue
> $ ./gradlew spotlessJavaApply
>
> Please ping me if you run into any difficulty.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/beam/pull/5545#issuecomment-401092921>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AUGE1ccwvUwiu5qovwBzcLVctY3A-2W7ks5uBQMdgaJpZM4UYOHO>
> .
>

kennknowles commented:
It only works during a rebase? You might need to provide files or `'**'` quoted?

kennknowles commented:
You really don't have to. I may have made some typo but it definitely works to simply discard one side or the other of a conflict.

apilloud commented on sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/FieldAccessDescriptor.java:
```diff
@@ -0,0 +1,265 @@
+    return fieldIds;
+  }
+
+  private Schema getFieldSchema(Field field) {
```
nit: This can be `static`?

reuvenlax replied:
done

apilloud commented on sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/FieldAccessDescriptor.java:
```diff
@@ -0,0 +1,265 @@
+    return null;
+  }
+
+  private FieldAccessDescriptor resolvedNestedFieldsHelper(
```
nit: This can be `static`?

apilloud commented on sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/FieldAccessDescriptor.java:
```diff
@@ -0,0 +1,265 @@
+    return nestedFields;
+  }
+
+  private int validateFieldId(Schema schema, int fieldId) {
```
nit: This can be `static`?

reuvenlax replied:
done

apilloud commented on sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/SchemaCoder.java:
```diff
@@ -0,0 +1,89 @@
+/** {@link SchemaCoder} is used as the coder for types that have schemas registered. */
+@Experimental(Kind.SCHEMAS)
+public class SchemaCoder<T> extends CustomCoder<T> {
+  private Schema schema;
```
This class also seems like a good candidate for AutoValue? If not, these should all be final.

reuvenlax replied:
feels less appropriate here; made variables final

apilloud commented on sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/SchemaRegistry.java:
```diff
@@ -0,0 +1,156 @@
+@Experimental(Kind.SCHEMAS)
+public class SchemaRegistry {
+  private static class SchemaEntry<T> {
+    Schema schema;
```
Generic `AutoValue` or `final` comment goes here.

reuvenlax replied:
AutoValue overkill here - made them final

apilloud commented on sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/SchemaRegistry.java:
```diff
@@ -0,0 +1,156 @@
+    }
+  }
+
+  Map<TypeDescriptor, SchemaEntry> entries = Maps.newHashMap();
```
These both are `final`?

reuvenlax replied:
done

reuvenlax commented:
retest this please

reuvenlax commented on sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/FieldAccessDescriptor.java:
```diff
@@ -0,0 +1,265 @@
+  }
+
+  public FieldAccessDescriptor resolve(Schema schema) {
+    Set<Integer> fieldIdsAccessed = resolveFieldIdsAccessed(schema);
```
Done

reuvenlax commented on sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/FieldAccessDescriptor.java:
```diff
@@ -0,0 +1,265 @@
+ */
+@Experimental(Kind.SCHEMAS)
+public class FieldAccessDescriptor implements Serializable {
+  private boolean allFields;
```
Made class AutoVAlue

reuvenlax commented:
retest this please

reuvenlax commented:
retest this please

reuvenlax commented:
run Dataflow PostCommit

reuvenlax commented:
Run Dataflow ValidatesRunner

reuvenlax commented:
Run Dataflow ValidatesRunner

reuvenlax commented:
Run Dataflow ValidatesRunner

lgajowy commented:
I'm afraid after merging this the build started breaking: 
https://builds.apache.org/view/A-D/view/Beam/job/beam_PostCommit_Java_GradleBuild/1002/console 

Better (more verbose) logs: https://builds.apache.org/view/A-D/view/Beam/job/beam_PerformanceTests_XmlIOIT/453/console

reuvenlax commented:
Looking. I explicitly triggered a bunch of PostCommit tests before merging
this, so might have been one I missed.

On Sat, Jun 30, 2018 at 11:03 AM Łukasz Gajowy <notifications@github.com>
wrote:

> I'm afraid after merging this the build started breaking:
>
> https://builds.apache.org/view/A-D/view/Beam/job/beam_PostCommit_Java_GradleBuild/1002/console
>
> Better (more verbose) logs:
> https://builds.apache.org/view/A-D/view/Beam/job/beam_PerformanceTests_XmlIOIT/453/console
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/beam/pull/5545#issuecomment-401556941>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AUGE1cNgHhxaKPNnKFaLMBSFlHmCI4bOks5uB71ygaJpZM4UYOHO>
> .
>

reuvenlax commented:
More specifically what happened: it appear that the conflicting PR was
merged after regular tests had triggered for this PR but before I had
merged my PR. The fact that I made sure to run ValidatesRunner tests before
merging my PR ironically caused this, as it widened the window between
tests running and PR merging (since it took many hours to run the
ValidatesRunner tests).

I sent you https://github.com/apache/beam/pull/5845 to fix this.

On Sat, Jun 30, 2018 at 11:25 AM Reuven Lax <relax@google.com> wrote:

> There is a compilation error, introduced by a conflict with
> https://github.com/apache/beam/commit/34edbd5cc38c1eab0307569da518bceef325e696#diff-6f8f4b5d3d2bc0ef3cc57759b224bc36.Sending
> you a PR to fix this.
>
> On Sat, Jun 30, 2018 at 11:10 AM Reuven Lax <relax@google.com> wrote:
>
>> Looking. I explicitly triggered a bunch of PostCommit tests before
>> merging this, so might have been one I missed.
>>
>> On Sat, Jun 30, 2018 at 11:03 AM Łukasz Gajowy <notifications@github.com>
>> wrote:
>>
>>> I'm afraid after merging this the build started breaking:
>>>
>>> https://builds.apache.org/view/A-D/view/Beam/job/beam_PostCommit_Java_GradleBuild/1002/console
>>>
>>> Better (more verbose) logs:
>>> https://builds.apache.org/view/A-D/view/Beam/job/beam_PerformanceTests_XmlIOIT/453/console
>>>
>>> —
>>> You are receiving this because you modified the open/close state.
>>> Reply to this email directly, view it on GitHub
>>> <https://github.com/apache/beam/pull/5545#issuecomment-401556941>, or mute
>>> the thread
>>> <https://github.com/notifications/unsubscribe-auth/AUGE1cNgHhxaKPNnKFaLMBSFlHmCI4bOks5uB71ygaJpZM4UYOHO>
>>> .
>>>
>>

reuvenlax commented:
Fix merged.

On Sat, Jun 30, 2018 at 11:32 AM Reuven Lax <relax@google.com> wrote:

> More specifically what happened: it appear that the conflicting PR was
> merged after regular tests had triggered for this PR but before I had
> merged my PR. The fact that I made sure to run ValidatesRunner tests before
> merging my PR ironically caused this, as it widened the window between
> tests running and PR merging (since it took many hours to run the
> ValidatesRunner tests).
>
> I sent you https://github.com/apache/beam/pull/5845 to fix this.
>
> On Sat, Jun 30, 2018 at 11:25 AM Reuven Lax <relax@google.com> wrote:
>
>> There is a compilation error, introduced by a conflict with
>> https://github.com/apache/beam/commit/34edbd5cc38c1eab0307569da518bceef325e696#diff-6f8f4b5d3d2bc0ef3cc57759b224bc36.Sending
>> you a PR to fix this.
>>
>> On Sat, Jun 30, 2018 at 11:10 AM Reuven Lax <relax@google.com> wrote:
>>
>>> Looking. I explicitly triggered a bunch of PostCommit tests before
>>> merging this, so might have been one I missed.
>>>
>>> On Sat, Jun 30, 2018 at 11:03 AM Łukasz Gajowy <notifications@github.com>
>>> wrote:
>>>
>>>> I'm afraid after merging this the build started breaking:
>>>>
>>>> https://builds.apache.org/view/A-D/view/Beam/job/beam_PostCommit_Java_GradleBuild/1002/console
>>>>
>>>> Better (more verbose) logs:
>>>> https://builds.apache.org/view/A-D/view/Beam/job/beam_PerformanceTests_XmlIOIT/453/console
>>>>
>>>> —
>>>> You are receiving this because you modified the open/close state.
>>>> Reply to this email directly, view it on GitHub
>>>> <https://github.com/apache/beam/pull/5545#issuecomment-401556941>, or mute
>>>> the thread
>>>> <https://github.com/notifications/unsubscribe-auth/AUGE1cNgHhxaKPNnKFaLMBSFlHmCI4bOks5uB71ygaJpZM4UYOHO>
>>>> .
>>>>
>>>
